src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/services/Services.java

```diff
@@ -261,22 +252,128 @@ protected void threadTerminated(T value) {
+        if (cstring == 0) {
+            return null;
+        }
+        int len = 0;
+        for (long p = cstring; unsafe.getByte(p) != 0; p++) {
+            len++;
+        }
+        byte[] buf = new byte[len];
+        for (int i = 0; i < len; i++) {
+            buf[i] = unsafe.getByte(cstring + i);
+        }
+        return new String(buf, java.nio.charset.StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Gets the value of {@code Arguments::systemProperties()} and puts the offsets
+     * of {@code SystemProperty} fields into {@code offsets}. The values returned in
+     * {@code offsets} are:
+     *
+     * <pre>
+     *     [ next,  // SystemProperty::next_offset_in_bytes()
+     *       key,   // SystemProperty::key_offset_in_bytes()
+     *       value  // PathString::value_offset_in_bytes()
+     *     ]
+     * </pre>
+     *
+     * Ideally this would be done with vmstructs but that code is in {@code jdk.vm.ci.hotspot}.
+     */
+    private static native long readSystemPropertiesInfo(int[] offsets);
+
     /**
-     * Initializes {@link #savedProperties} from the byte array returned by
-     * {@code jdk.internal.vm.VMSupport.serializeSavedPropertiesToByteArray()}.
+     * Parses the native {@code Arguments::systemProperties()} data structure using Unsafe to
+     * create a properties map. This parsing is safe as argument parsing in completed in
+     * early VM start before this code can be executed, making {@code Arguments::systemProperties()}
+     * effectively read-only by now.
      */
-    @VMEntryPoint
-    private static void initializeSavedProperties(byte[] serializedProperties) throws IOException {
-        if (!IS_IN_NATIVE_IMAGE) {
-            throw new InternalError("Can only initialize saved properties in JVMCI shared library runtime");
+    private static Map<String, String> initProperties() {
+        int[] offsets = new int[3];
+        long systemProperties = readSystemPropertiesInfo(offsets);
+        int nextOffset = offsets[0];
+        int keyOffset = offsets[1];
+        int valueOffset = offsets[2];
+
+        int count = 0;
+        Unsafe unsafe = Unsafe.getUnsafe();
+        for (long prop = systemProperties; prop != 0; prop = unsafe.getLong(prop + nextOffset)) {
+            if (unsafe.getLong(prop + valueOffset) != 0) {
+                count++;
+            } else {
+                // Some internal properties (e.g. jdk.boot.class.path.append) can have a null
+                // value and should just be ignored. Note that null is different than the empty string.
+            }
+        }
+        Map<String, SystemProperties.Value> props = new HashMap<>(count + 1);
+        int i = 0;
+        for (long prop = systemProperties; prop != 0; prop = unsafe.getLong(prop + nextOffset)) {
+            String key = toJavaString(unsafe, unsafe.getLong(prop + keyOffset));
+            long valueAddress = unsafe.getLong(prop + valueOffset);
+            if (valueAddress != 0) {
```

Comment:
Is the lazy value construction really necessary?  It seems like it requires quite a bit of extra machinery.